### PR TITLE
gh-117657: Fix TSan reported data race on ioctl_works

### PR DESCRIPTION
**Pull request by colesbury**
* Issue: gh-117657

### CODE REVIEW

**Review by corona10 — APPROVED** [review]

**Review by vstinner — APPROVED** [review]
LGTM.

I tested functionally that setting ioctl_works to 0 falls back on fcntl() as expected by forcing ioctl_works to 0 and by checking syscalls with strace.

File: Python/fileutils.c [hunk 1]
```diff
@@ -1512,7 +1512,9 @@ set_inheritable(int fd, int inheritable, int raise, int *atomic_flag_works)
             request = FIOCLEX;
         err = ioctl(fd, request, NULL);
         if (!err) {
-            ioctl_works = 1;
+            if (_Py_atomic_load_int_relaxed(&ioctl_works) == -1) {
```
Comment (aisk) [inline Python/fileutils.c:1515]: Should we use the atomic operations on non free-threads build?

Comment (colesbury) [reply Python/fileutils.c:1515]: Yes, it's fine. There is no performance implication -- these relaxed load/stores get compiled to the same instructions as non-atomic operations.

Comment (aisk) [reply Python/fileutils.c:1515]: Got it, thanks for the explanation!

**Pull request merged**